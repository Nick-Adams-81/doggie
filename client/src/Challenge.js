import React, { useState } from "react";
import ItemRow from "./ItemRow";

export default function Challenge() {
  const [lowStock, setLowStock] = useState({
    data: "",
    loading: true,
  });
  const [restockCost, setRestockCost] = useState([]);

  const handleClickLowStock = async () => {
    try {
      const data = await fetch("http://localhost:4567/low-stock");
      const lowStock = await data.json();
      setLowStock({
        data: lowStock,
        loading: false,
      });
      return lowStock;
    } catch (e) {
      return null;
    }
  };

  const handleClickReorderCost = async () => {
    try {
      const data = await fetch("http://localhost:4567/restock-cost", {
        method: "post",
        headers: { "Content-type": "application/json" },
        body: JSON.stringify(lowStock.data),
      });
      const lowStockData = await data.json();
      setRestockCost(lowStockData);
    } catch (err) {
      return null;
    }
  };

  return (
    <>
      <table>
        <thead>
          <tr>
            <td data-testid="sku-text">SKU</td>
            <td>Item Name</td>
            <td>Amount in Stock</td>
            <td>Capacity</td>
            <td>Order Amount</td>
          </tr>
        </thead>
        <tbody>
          {lowStock.loading
            ? "loading..."
            : lowStock.data.map((item) => (
                <ItemRow
                  key={item.id}
                  item={item}
                  lowStock={lowStock}
                  setLowStock={setLowStock}
                />
              ))}
        </tbody>
      </table>
      <div id="totalCost" data-testid="total-cost">
        Total Cost: ${Math.round((restockCost + Number.EPSILON) * 100) / 100}
      </div>
      <button
        id="low-stock-button"
        data-testid="low-stock-button"
        onClick={handleClickLowStock}
      >
        Get Low-Stock Items
      </button>
      <button
        id="re-order-button"
        data-testid="re-order-button"
        onClick={handleClickReorderCost}
      >
        Determine Re-Order Cost
      </button>
    </>
  );
}
