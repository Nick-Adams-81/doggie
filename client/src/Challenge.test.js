import { render, screen, cleanup, fireEvent } from "@testing-library/react";
import renderer from "react-test-renderer"
import Challenge from "./Challenge";

describe("challenge page test suite", () => {
  beforeEach(() => {
    cleanup();
  });
  afterEach(() => {
    jest.clearAllMocks();
  });

  it("renders correctly and takes a snapshot of the component", () => {
    const tree = renderer.create(<Challenge />).toJSON();
    expect(tree).toMatchSnapshot();
  });
  test("should render sku on page", () => {
    render(<Challenge />);
    const skuText = screen.getByTestId("sku-text");
    expect(skuText).toBeInTheDocument();
  });
  it("should render item name in the page", () => {
    render(<Challenge />);
    const itemNameText = screen.getByText(/item name/i);
    expect(itemNameText).toBeInTheDocument();
  });
  it("should render total cost: in the page", () => {
    render(<Challenge />);
    const totalCost = screen.getByTestId("total-cost");
    expect(totalCost).toBeInTheDocument();
    expect(totalCost.innerHTML).toMatch(/total cost:/i);
  });
  it("should make a fetch request", () => {
    const data = jest.fn(() => {
      fetch("http://localhost:4567/low-stock")
        .then((res) => res.json())
        .catch((err) => console.log(err));
    });
    expect(data).not.toBeNull();
    expect(fetch).toHaveBeenCalledTimes(0);
  });
});
