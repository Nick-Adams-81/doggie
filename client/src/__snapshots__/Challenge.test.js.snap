// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`challenge page test suite renders correctly and takes a snapshot of the component 1`] = `
Array [
  <table>
    <thead>
      <tr>
        <td
          data-testid="sku-text"
        >
          SKU
        </td>
        <td>
          Item Name
        </td>
        <td>
          Amount in Stock
        </td>
        <td>
          Capacity
        </td>
        <td>
          Order Amount
        </td>
      </tr>
    </thead>
    <tbody>
      loading...
    </tbody>
  </table>,
  <div
    data-testid="total-cost"
    id="totalCost"
  >
    Total Cost: $
    0
  </div>,
  <button
    data-testid="low-stock-button"
    id="low-stock-button"
    onClick={[Function]}
  >
    Get Low-Stock Items
  </button>,
  <button
    data-testid="re-order-button"
    id="re-order-button"
    onClick={[Function]}
  >
    Determine Re-Order Cost
  </button>,
]
`;
